oauth2_http/javatests/com/google/auth/oauth2/ITWorkloadIdentityFederationTest.java

```diff
@@ -0,0 +1,261 @@
+import com.google.api.client.http.javanet.NetHttpTransport;
+import com.google.api.client.json.GenericJson;
+import com.google.api.client.json.JsonFactory;
+import com.google.api.client.json.JsonObjectParser;
+import com.google.api.client.json.gson.GsonFactory;
+import com.google.api.client.util.GenericData;
+import com.google.auth.http.HttpCredentialsAdapter;
+import java.io.ByteArrayInputStream;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.contrib.java.lang.system.EnvironmentVariables;
+
+/**
+ * Integration tests for Workload Identity Federation.
+ *
+ * <p>The only requirement for this test suite to run is to set the environment variable
+ * GOOGLE_APPLICATION_CREDENTIALS to point to the same service account keys used in the setup script
+ * (workloadidentityfederation-setup).
+ *
+ * <p>1. IdentityPoolCredentials (OIDC provider): Uses the service account to generate a Google ID
+ * token using the iamcredentials generateIdToken API. This will use the service account client ID
+ * as the sub field of the token. This OIDC token will be used as the external subject token to be
+ * exchanged for a GCP access token via GCP STS endpoint and then to impersonate the original
+ * service account key.
+ *
+ * <p>2. AwsCredentials (AWS provider): Uses the service account keys to generate a Google ID token
+ * using the iamcredentials generateIdToken API. Exchanges the OIDC ID token for AWS security keys
+ * using AWS STS AssumeRoleWithWebIdentity API. These values will be set as AWS environment
+ * variables to simulate an AWS VM. The Auth library can now read these variables and create a
+ * signed request to AWS GetCallerIdentity. This will be used as the external subject token to be
+ * exchanged for a GCP access token via GCP STS endpoint and then to impersonate the original
+ * service account key.
+ */
+public final class ITWorkloadIdentityFederationTest {
+
+  // Copy output from workloadidentityfederation-setup.
+  private static final String AUDIENCE_PREFIX =
+      "//iam.googleapis.com/projects/1016721519174/locations/global/workloadIdentityPools/pool-1/providers/";
+  private static final String AWS_ROLE_NAME = "ci-java-test";
+  private static final String AWS_ROLE_ARN = "arn:aws:iam::027472800722:role/ci-java-test";
+
+  private static final String AWS_AUDIENCE = AUDIENCE_PREFIX + "aws-1";
+  private static final String OIDC_AUDIENCE = AUDIENCE_PREFIX + "oidc-1";
+
+  private String clientEmail;
+
+  @Rule public final EnvironmentVariables environmentVariables = new EnvironmentVariables();
+
+  @Before
+  public void setup() throws IOException {
+    GenericJson keys = getServiceAccountKeyFileAsJson();
+    clientEmail = (String) keys.get("client_email");
+  }
+
+  @Test
+  public void identityPoolCredentials() throws IOException {
```

Comment:
nit: can the test name better describe what it's trying to test

Reply:
I can't think of a name that accurately captures what is being tested. I did write a long comment above describing both tests though - do you want me to move it here?

Reply:
Yeah, it's probably better to have those next to the individual tests.